src/ispypsa/translator/generators.py

```diff
@@ -1,60 +1,998 @@
+    _ECAA_GENERATOR_ATTRIBUTES,
+    _GENERATOR_ATTRIBUTE_ORDER,
+    _NEW_ENTRANT_GENERATOR_ATTRIBUTES,
+)
 from ispypsa.translator.temporal_filters import _time_series_filter
 from ispypsa.translator.time_series_checker import _check_time_series
 
 
 def _translate_ecaa_generators(
-    ecaa_generators: pd.DataFrame, regional_granularity: str = "sub_regions"
+    ispypsa_tables: dict[str, pd.DataFrame],
+    investment_periods: list[int],
+    regional_granularity: str = "sub_regions",
+    rez_handling: str = "discrete_nodes",
+    year_type: str = "fy",
 ) -> pd.DataFrame:
     """Process data on existing, committed, anticipated, and additional (ECAA) generators
     into a format aligned with PyPSA inputs.
 
     Args:
-        ecaa_generators: `ISPyPSA` formatted pd.DataFrame detailing the ECAA generators.
+        ispypsa_tables: dictionary of dataframes providing the `ISPyPSA` input tables.
+            (add link to ispypsa input tables docs).
+        investment_periods: list of years in which investment periods start obtained
+            from the model configuration.
         regional_granularity: Regional granularity of the nodes obtained from the model
             configuration. Defaults to "sub_regions".
+        year_type: str which should be "fy" or "calendar". If "fy" then investment
+            period ints are interpreted as specifying financial years (according to the
+            calendar year the financial year ends in).
+        rez_handling: str from the model configuration that defines whether REZs are
+            modelled as distinct nodes, one of "discrete_nodes" or "attached_to_parent_node". Defaults to "discrete_nodes".
 
     Returns:
-        `pd.DataFrame`: `PyPSA` style generator attributes in tabular format.
+        `pd.DataFrame`: `PyPSA` style ECAA generator attributes in tabular format.
     """
 
-    gen_attributes = _GENERATOR_ATTRIBUTES.copy()
+    ecaa_generators = ispypsa_tables["ecaa_generators"]
+    if ecaa_generators.empty:
+        # TODO: log
+        # raise error?
+        return pd.DataFrame()
+
+    # calculate lifetime based on expected closure_year - build_year:
+    ecaa_generators = _add_closure_year_column(
+        ecaa_generators, ispypsa_tables["closure_years"], investment_periods
+    )
+    ecaa_generators["lifetime"] = ecaa_generators["closure_year"].map(
+        lambda x: float(x - investment_periods[0]) if x > 0 else np.inf
+    )
+    ecaa_generators = ecaa_generators[ecaa_generators["lifetime"] > 0].copy()
 
+    gen_attributes = _ECAA_GENERATOR_ATTRIBUTES.copy()
     if regional_granularity == "sub_regions":
         gen_attributes["sub_region_id"] = "bus"
+        bus_column = "sub_region_id"
     elif regional_granularity == "nem_regions":
         gen_attributes["region_id"] = "bus"
+        bus_column = "region_id"
+
+    if regional_granularity == "single_region":
+        ecaa_generators["bus"] = "NEM"
+        gen_attributes["bus"] = "bus"
+        bus_column = "bus"
```

Comment:
This is really trival, but maybe this is a bit clearer?

```suggestion

    # Decide which column to rename to be the bus column.
    if regional_granularity == "sub_regions":
        bus_column = "sub_region_id"
    elif regional_granularity == "nem_regions":
        bus_column = "region_id"
    elif regional_granularity == "single_region":
        # No existing column to use for bus, so create a new one.
        ecaa_generators["bus"] = "NEM"
        bus_column = "bus" # Name doesn't need to change.
    gen_attributes[bus_column] = "bus"
```